apps/prairielearn/src/lib/file-store.ts

```diff
@@ -120,79 +131,82 @@ export async function deleteFile(file_id, authn_user_id) {
  */
-export async function getFile(file_id, data_type = 'buffer') {
-  debug(`get(): file_id=${file_id}`);
-  const params = { file_id };
-  const result = await sqldb.queryZeroOrOneRowAsync(sql.select_file, params);
-  debug('get(): got row from DB');
-
-  let buffer, readStream;
-
-  if (result.rows.length < 1) {
+export async function getFile(
+  file_id: number | string,
+  data_type: 'stream' | 'buffer' = 'buffer',
+): Promise<{
+  contents: Buffer | Stream;
+  file: File;
+}> {
+  const file = await sqldb.queryOptionalRow(sql.select_file, { file_id }, FileSchema);
+
+  if (!file) {
     throw new Error(`No file with file_id ${file_id}`);
   }
 
-  if (result.rows[0].storage_type === StorageTypes.FileSystem) {
+  if (file.storage_type === StorageTypes.FileSystem) {
     if (config.filesRoot == null) {
       throw new Error(
         `config.filesRoot must be non-null to get file_id ${file_id} from file store`,
       );
     }
 
-    const filename = path.join(config.filesRoot, result.rows[0].storage_filename);
+    const filename = path.join(config.filesRoot, file.storage_filename);
 
-    if (data_type === 'buffer') {
-      debug(`get(): readFile ${filename} and return object with contents buffer and file object`);
-      buffer = await fsPromises.readFile(filename);
-    } else {
-      debug(
-        `get(): createReadStream ${filename} and return object with contents stream and file object`,
-      );
-      readStream = fs.createReadStream(filename);
-    }
-  }
+    const contents =
+      data_type === 'buffer' ? await fsPromises.readFile(filename) : fs.createReadStream(filename);
 
-  if (result.rows[0].storage_type === StorageTypes.S3) {
+    return {
+      contents,
+      file,
+    };
+  } else if (file.storage_type === StorageTypes.S3) {
     if (config.fileStoreS3Bucket == null) {
       throw new Error(
-        `config.fileStoreS3Bucket must be configured to get file_id ${file_id} from file store`,
+        `config.fileStoreS3Bucket must be non-null to get file_id ${file_id} from file store`,
       );
     }
 
-    if (data_type === 'buffer') {
-      debug(
-        `get(): s3 fetch file ${result.rows[0].storage_filename} from ${config.fileStoreS3Bucket} and return object with contents buffer and file object`,
-      );
-      buffer = await getFromS3(config.fileStoreS3Bucket, result.rows[0].storage_filename);
-    } else {
-      debug(
-        `get(): s3 fetch stream ${result.rows[0].storage_filename} from ${config.fileStoreS3Bucket} and return object with contents stream and file object`,
-      );
-      readStream = await getFromS3(
-        config.fileStoreS3Bucket,
-        result.rows[0].storage_filename,
-        false,
-      );
-    }
-  }
+    const contents =
+      data_type === 'buffer'
+        ? await getFromS3(config.fileStoreS3Bucket, file.storage_filename, true)
+        : await getFromS3(config.fileStoreS3Bucket, file.storage_filename, false);
```

Comment:
I would have tried to write this as follows, but it turns out TypeScript doesn't like this!

```suggestion
      await getFromS3(config.fileStoreS3Bucket, file.storage_filename, data_type === 'buffer')
```

I'm not sold on having a single function that returns different types depending on parameters. `getBufferFromS3` and `getStreamFromS3` would be nice (same suggestion as `getFileBuffer` and `getFileStream` in this file).

I won't block on this.